src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -33,140 +32,192 @@ internal partial class SyntacticClassificationTaggerProvider
+            {
+                var token = _listener.BeginAsyncOperation(nameof(OnWorkspaceRegistrationChanged));
+                var task = SwitchToMainThreadAndHookupWorkspaceAsync();
+                task.CompletesAsyncOperation(token);
+            }
+
+            private async Task SwitchToMainThreadAndHookupWorkspaceAsync()
             {
-                // We both try to connect synchronously, and register for workspace registration events.
-                // It's possible (particularly in tests), to connect in the startup path, but then get a
-                // previously scheduled, but not yet delivered event.  Don't bother connecting to the
-                // same workspace again in that case.
-                var newWorkspace = _workspaceRegistration.Workspace;
-                if (newWorkspace == _workspace)
+                try
                 {
-                    return;
-                }
+                    await this.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(_disposalCancellationSource.Token);
+
+                    // We both try to connect synchronously, and register for workspace registration events.
+                    // It's possible (particularly in tests), to connect in the startup path, but then get a
+                    // previously scheduled, but not yet delivered event.  Don't bother connecting to the
+                    // same workspace again in that case.
+                    var newWorkspace = _workspaceRegistration.Workspace;
+                    if (newWorkspace == _workspace)
+                        return;
 
-                DisconnectFromWorkspace();
+                    DisconnectFromWorkspace();
 
-                if (newWorkspace != null)
+                    if (newWorkspace != null)
+                        ConnectToWorkspace(newWorkspace);
+                }
+                catch (OperationCanceledException)
                 {
-                    ConnectToWorkspace(newWorkspace);
+                    // can happen if we were disposed of.
+                }
+                catch (Exception e) when (FatalError.ReportAndCatch(e))
+                {
+                    // We were in a fire and forget task.  So just report the NFW and do not let
+                    // this bleed out to an unhandled exception.
                 }
             }
 
             internal void IncrementReferenceCount()
-                => _taggerReferenceCount++;
+            {
+                this.AssertIsForeground();
+                _taggerReferenceCount++;
+            }
 
-            internal void DecrementReferenceCountAndDisposeIfNecessary()
+            internal void DecrementReferenceCount()
             {
+                this.AssertIsForeground();
                 _taggerReferenceCount--;
 
                 if (_taggerReferenceCount == 0)
                 {
+                    // stop any bg work we're doing.
+                    _disposalCancellationSource.Cancel();
```

Comment:
the only cancellation we have is for when we actually get disposed once hte last buffer throws us away.  all other work always proceeds to completion.